DeviceAdministration/Infrastructure.UnitTests/Web/WebApiControllers/TelemetryApiControllerTests.cs

```diff
@@ -0,0 +1,191 @@
+            this.fixture.Create<IEnumerable<DeviceTelemetryModel>>();
+            var telemetryFields = this.fixture.Create<IList<DeviceTelemetryFieldModel>>();
+
+            this.deviceLogic.Setup(mock => mock.GetDeviceAsync(deviceId)).ReturnsAsync(device);
+            this.deviceLogic.Setup(mock => mock.ExtractTelemetry(device)).Returns(telemetryFields);
+            this.telemetryLogic.Setup(mock => mock.LoadLatestDeviceTelemetryAsync(deviceId, telemetryFields, minTime)).ReturnsAsync(telemetryModel);
+
+            var res = await this.telemetryApiController.GetDeviceTelemetryAsync(deviceId, minTime);
+            res.AssertOnError();
+            var data = res.ExtractContentAs<DeviceTelemetryModel[]>();
+            Assert.Equal(data, telemetryModel.OrderBy(t => t.Timestamp).ToArray());
+
+            this.deviceLogic.Setup(mock => mock.ExtractTelemetry(device)).Throws(new Exception());
+            await Assert.ThrowsAsync<HttpResponseException>(() => this.telemetryApiController.GetDeviceTelemetryAsync(deviceId, minTime));
+            res.AssertOnError();
+        }
+
+        [Fact]
+        public async void GetDeviceTelemetrySummaryAsyncTest()
+        {
+            var telemetrySummary = this.fixture.Create<DeviceTelemetrySummaryModel>();
+            var deviceId = this.fixture.Create<string>();
+            this.telemetryLogic.Setup(mock => mock.LoadLatestDeviceTelemetrySummaryAsync(deviceId, null)).ReturnsAsync(telemetrySummary);
+            var res = await this.telemetryApiController.GetDeviceTelemetrySummaryAsync(deviceId);
+            res.AssertOnError();
+            var data = res.ExtractContentAs<DeviceTelemetrySummaryModel>();
+            Assert.Equal(data, telemetrySummary);
+        }
+
+        [Fact]
+        public async void GetLatestAlertHistoryAsyncTest()
+        {
+            var result = this.fixture.Create<DeviceListQueryResult>();
+            var locationsModel = this.fixture.Create<DeviceListLocationsModel>();
+            var itemModels = this.fixture.Create<IEnumerable<AlertHistoryItemModel>>();
+
+            this.alertLogic.Setup(mock => mock.LoadLatestAlertHistoryAsync(It.IsAny<DateTime>(), It.IsAny<int>())).ReturnsAsync(itemModels);
+            this.deviceLogic.Setup(mock => mock.GetDevices(It.IsAny<DeviceListQuery>())).ReturnsAsync(result);
+            this.deviceLogic.Setup(mock => mock.ExtractLocationsData(It.IsAny<List<DeviceModel>>())).Returns(locationsModel);
+            this.telemetryLogic.Setup(mock => mock.ProduceGetLatestDeviceAlertTime(It.IsAny<IEnumerable<AlertHistoryItemModel>>()))
+                .Returns(this.GetStatusTimeNull);
+
+            var res = await this.telemetryApiController.GetLatestAlertHistoryAsync();
+            res.AssertOnError();
+            var data = res.ExtractContentAs<AlertHistoryResultsModel>();
+            if (itemModels.Count() < 18)
+                Assert.Equal(data.Data.Count, itemModels.Count());
```

Comment:
it's not really a if/else. it's more of take the first 18 elements:
`resultsModel.Data = historyItems.Take(DISPLAYED_HISTORY_ITEMS).ToList();`